README.md
# MLstats

Deep learning and machine learning algorithms with specific likelihoods


[![Build Status](https://travis-ci.org/MaximilianPi/MLstats.svg?branch=master)](https://travis-ci.org/MaximilianPi/MLstats) 

install development version:
 ```{r}
devtools::install_github(repo = "MaximilianPi/MLstats",subdir = "MLstats", build_vignettes = T)
```


MLstats/NEWS.md
# MLstats 0.1.0

* Added a `NEWS.md` file to track changes to the package.


MLstats/tests/testthat.R
library(testthat)
library(MLstats)

test_check("MLstats")


MLstats/tests/testthat/test-lm.R
testthat::context("lm")

skip_if_no_tensorflow <- function() {
  if (!reticulate::py_module_available("tensorflow"))
    skip("TensorFlow not available for testing")
}

x = matrix(runif(1000), nrow = 100, ncol = 10)
w = runif(10,-1,1)
y = x %*% w + 2 + rnorm(100,0,0.5)
data = data.frame(x, y=y)
# parameter = NULL


testthat::test_that("lm functionality",{
  testthat::expect_error(lmML("egal"))
  testthat::expect_error(lmML())
  testthat::expect_error(lmML(y~., data = data, parameter = list(hi = "hi")))
  testthat::expect_error(lmML(y~., data = data, parameter = list(hi = "hi"), method = "rf"))
  testthat::expect_error(lmML(y~., data = data, parameter = list(epochs = "hi"), method = "ann"))

  skip_if_no_tensorflow()
  testthat::expect_error(suppressWarnings(lmML(y~., data = data, parameter = list(epochs = 1L, architecture = 1L), method = "ann")), NA)
  testthat::expect_error(suppressWarnings({
    fit = lmML(y~., data = data, parameter = list(epochs = 1L, architecture = 1L), method = "ann")
    p = predict(fit,data)
    }), NA)
  testthat::expect_error(suppressWarnings({
    fit = lmML(y~., data = data, parameter = list(epochs = 1L, architecture = 1L), method = "ann")
    p = predict(fit,data)
  }), NA)
})


MLstats/tests/testthat/test-ann.R
testthat::context("ann functionality")
skip_if_no_tensorflow <- function() {
  if (!reticulate::py_module_available("tensorflow"))
    skip("TensorFlow not available for testing")
}

x = matrix(runif(1000), nrow = 100, ncol = 10)
w = runif(10,-1,1)
y = x %*% w + 2 + rnorm(100,0,0.5)
data = list(x=x, y=y)
# parameter = NULL
# formula = y~.
# cv = NULL
# method = "ann"
testthat::test_that("ann parameters", {
  skip_if_no_tensorflow()
  testthat::expect_error({ANN$new()})
  testthat::expect_error({ANN$new(data = data, batch_size = 20)}, NA)
  testthat::expect_error({ANN$new(data = data, epochs = 0)})
  testthat::expect_error({ANN$new(data = data, learning_rate = -1)})
  testthat::expect_error({ANN$new(data = data, dropout_rate  = -1)})
  testthat::expect_error({ANN$new(data = data, architecture = NULL)})
  testthat::expect_error({ANN$new(data = data, architecture = 10L)}, NA)
  testthat::expect_error({ANN$new(data = data, activation_function = "egal")})
})


testthat::test_that("ann functionality", {
  skip_if_no_tensorflow()
  testthat::expect_error({
    m = ANN$new(data = data, epochs = 1L, architecture = 1L)
    m$build()
    m$compile()
    m$train()
  }, NA)

  testthat::expect_error({
    m = ANN$new(data = data, epochs = 1L, architecture = 1L)
    m$build()
    m$compile()
    m$train()
    m$predict()
  }, NA)

  testthat::expect_error({
    m = ANN$new(data = data, epochs = 1L, architecture = 1L)
    m$build()
    m$compile()
    m$train()
    m$predict(data$x)
  }, NA)

  testthat::expect_error({
    m = ANN$new(data = data, epochs = 1L, architecture = 1L)
    m$build()
    m$compile()
    m$train()
    m$reset_fit()
    m$train()
  }, NA)

  testthat::expect_error({
    m = ANN$new(data = data, epochs = 1L, architecture = 1L)
    m$build()
    m$compile()
    m$train()
    m$reset_all()
    m$train()
  })

  testthat::expect_error({
    m = ANN$new(data = data, epochs = 1L, architecture = 1L)
    m$build()
    m$compile()
    m$train()
    m$reset_all()
    m$build()
    m$compile()
    m$train()
  }, NA)
})



MLstats/R/lm.R
#'Fit a linear regression with a normal distributed error
#'@author Maximilian Pichler
#'
#'@param  formula an object of class formula
#'@param data data frame
#'@param subset indicies on which model is fitted
#'@param na.action how to handle nas
#'@param method support methods are random forest, deep neural networks
#'@param scale how to handle unscales values
#'@param parameter list of parameter for method
#'@param cv cross validation strategy
#'
#'@importFrom stats model.frame
#'@importFrom stats model.matrix
#'@importFrom stats model.response
#'@importFrom stats dnorm
#'@export

# TO DO: implement cv + subset + na.action +examples+description

lmML = function(formula, data = NULL, subset = NULL, na.action = NULL, method = "ann", scale = FALSE, parameter = NULL, cv = NULL){

  if(!inherits(formula, "formula")) stop("method is only for formula objects")
  if(method != "ann") stop("At the moment, only method = 'ann' is supported")
  if(is.null(parameter)) parameter = get_default_parameter(method = method)
  parameter = check_parameter(parameter, method)
  if(is.logical(parameter) && !parameter) stop("Unkown parameters in parameter list!")

  # not elegant...change later
  # ......................... #
  y = model.response(model.frame(formula, data))

  x = model.matrix(formula, data)

  x = x[,!colnames(x) %in% "(Intercept)"]
  # ......................... #

  data_prepared = list(x = x, y = y)

  model = create_model_object(method, parameter)

  model_fitted = model_fit(model, cv, data_prepared)

  predictions = predict(model_fitted, data_prepared$x)

  residuals = data_prepared$y - predictions

  output = list()

  output$predictions = predictions

  output$residuals = residuals

  output$deviance = -2*sum(dnorm(data_prepared$y, predictions, sd = model_fitted$lm$sigma, log = T))

  output$model = model_fitted$lm

  output$formula = formula

  class(output) = paste0("MLstats.", class(model_fitted))

  return(output)

  #importance = get_importance(model_fit)

}


#' Create model object for useage of methods
#'
#' @author Maximilian Pichler
#' @param method default ann
#' @param parameter parameter list

create_model_object = function(method, parameter){
  out = list()
  if(method == "ann"){
    out$model = method
    out$parameter = parameter
    class(out) = "lm_ann"
    return(out)
  }
}



#' check parameter list
#' @author Maximilian Pichler
#' @param parameter parameter list to check
#' @param method method, only ann supported

check_parameter = function(parameter, method){
  get_default = get_default_parameter()
  if(!any(names(parameter) %in% names(get_default))) return(FALSE)
  else {
    for(p in names(parameter)) get_default[[p]] = parameter[[p]]
    return(get_default)
  }
}





#' Get default parameter function, internal
#' @author MaximilianPi
#' @param method pars for method, default ann
get_default_parameter = function(method = "ann") {
  if(method == "ann"){
    out = list()
    out$architecture = c(10,10)
    out$lr = 0.001
    out$regularization = "batch_normalization"
    out$activation_function = "relu"
    out$batch = 25L
    out$epochs = 100L
    out$learning_rate = 0.001
    out$bias = TRUE
    out$dropout_rate = 1.
  }
  return(out)
}


#' Model predict function for lm_ann
#'
#' @author Maximilian Pichler
#' @param model model type
#' @param data predict for data
#' @importFrom stats predict
#' @export
predict.lm_ann = function(model, data){
  return(model$lm$predict(data))
}


#' generics test
#' @param model model..
#' @param data data...
#' @export

predict.MLstats = function(model,data){
  UseMethod("predict.MLstats", model)
}

#' Model predict function for MLstats.lm_ann
#'
#' @author Maximilian Pichler
#' @param model model type
#' @param data predict for data
#' @importFrom stats predict
#' @importFrom stats model.response
#' @importFrom stats model.matrix
#' @importFrom stats model.frame
#' @export
predict.MLstats.lm_ann = function(model, data){

  y = model.response(model.frame(model$formula, data))

  x = model.matrix(model$formula, data)

  x = x[,!colnames(x) %in% "(Intercept)"]

  return(model$model$predict(as.matrix(x)))
}

#' Model fit function for the different models
#'
#' @author Maximilian Pichler
#' @param model model type
#' @param cv type
#' @param data data for fitting
#' @export
model_fit = function(model, cv, data){
  UseMethod("model_fit", model)
}


#' Model fit lm for ANN
#'
#' @author Maximilian Pichler
#' @param model model type
#' @param cv cv type
#' @param data data..
#' @export
model_fit.lm_ann = function(model, cv, data){

  parameter = model$parameter

  model$lm = ANN_lm$new(architecture = parameter$architecture,
                        activation_function = parameter$activation_function,
                        bias = parameter$bias,
                        regularization = parameter$regularization,
                        batch_size = parameter$batch,
                        epochs = parameter$epochs,
                        learning_rate = parameter$learning_rate,
                        data = data)

  model$lm$build_lm()
  model$lm$compile()
  model$lm$train()
  return(model)

}


#' Model fit lm for ANN
#'
#' @author Maximilian Pichler
#' @param model model type
#' @param cv cv type
#' @param data data ...
#' @export
model_fit.lm_rf = function(model, cv, data){

}


MLstats/R/ann.R
#' Base ANN object
#'
#' \code{ANN} object holds a basic artifical neural network model
#'
#' @format \code{\link{R6Class}} object.
#' @name ANN
#' @section Usage:
#' \preformatted{
#' model = ANN$new()
#' }
#'
#' @section Arguments:
#' \describe{
#' \item{model: }{(any)\cr keras model}
#' \item{data: }{(list)\cr List of data with x and y as matrices }
#' \item{data: }{(list)\cr fit history list}
#' \item{build: }{(function)\cr build function}
#'
#'
#' }
#'
#' @section Details:
#' comming soon
#'
#'
#'
#' @author Maximilian Pichler
#' @export
NULL

#'@export
ANN = R6::R6Class("ANN",
 public = list(
   model = NULL,
   data = NULL,
   fit_history = NULL,
   k = NULL,
   initialize = function(architecture = c(10,10),
                         activation_function = "relu",
                         bias = TRUE,
                         regularization = "batch_normalization",
                         dropout_rate = 1.,
                         batch_size = 25L,
                         epochs = 100L,
                         learning_rate = 0.001,
                         data = NULL){
     self$k = keras::backend()
     self$reset_all()

     if(length(architecture) <= 0) stop("For linear regression, switch to lm()",call. = FALSE)
     if(!is.numeric(architecture)) stop("Numerical vector for architecture is needed", call. = FALSE)
     if(any(sapply(architecture, function(x) x == 0))) stop("Nodes with 0 units not allowed", call. = FALSE)
     if(!activation_function %in% c("relu", "tanh", "elu")) stop("activation_function must be one of (relu, tanh, elu)", call. = FALSE)
     if(dropout_rate > 1 || dropout_rate < 0) stop("dropout_rate must be in [0,1]", call. = FALSE)
     if(batch_size > nrow(data$x)) {
       batch_size = nrow(data$x)
       warning("Setting batch_size to nrow(data) because batch size cannot be larger than data", call. = FALSE)
     }
     if(epochs <=0 || !is.numeric(epochs)) stop("Epochs must be > 0 and numeric", call. = FALSE)
     if(learning_rate <= 0 || !is.numeric(learning_rate)) stop("learning_rate must be > 0 and numeric", call. = FALSE)
     if(is.null(data)) warning("Data must contain data or model cannot be built and trained", call. = FALSE)

     private$architecture = architecture
     private$activation_function = activation_function
     private$bias = bias
     private$regularization = regularization
     private$dropout_rate = dropout_rate
     private$batch_size = batch_size
     private$epochs = epochs
     private$learning_rate = learning_rate

     if(is.null(dim(data$y))) {
       warning("Y is a vector, changing to matrix with ncol == 1",call. = FALSE)
       data$y = matrix(data$y, ncol = 1, byrow = TRUE)
     }

     self$data = data
     self$model = keras::keras_model_sequential()
     private$built = FALSE

   },
   build = function(){
     if(!private$built){
       regularization =
         if(private$regularization=="batch_normalization") {
           function(model) return(keras::layer_batch_normalization(model))
         } else {
           function(model) return(keras::layer_dropout(model, rate = self$dropout_rate))
         }

       for(i in 1:length(private$architecture)){
         if(i == 1) {
           keras::layer_dense(self$model, input_shape = dim(self$data$x[2]),
                              activation = private$activation_function, units = private$architecture[i])
         }else {
           keras::layer_dense(self$model,
                              activation = private$activation_function, units = private$architecture[i])
         }
         regularization(self$model)
       }

       private$built = TRUE
       return(TRUE)
     } else {
       warning("already built, reset first", call. = FALSE)
       return(FALSE)
     }
   },

   build_output = function(output_activation = "relu"){
       keras::layer_dense(self$model, units = dim(self$data$y)[2], activation = output_activation)
   },

   compile = function(){
     keras::compile(self$model, loss = keras::loss_mean_squared_error, optimizer = keras::optimizer_rmsprop(lr = private$learning_rate))
   },

   train = function(){
     fit_history = keras::fit(self$model, x = self$data$x, y = self$data$y, epochs = private$epochs, batch_size = private$batch_size, shuffle = TRUE)
     self$fit_history = c(self$fit_history, fit_history$metrics$loss)
   },

   predict = function(newdata = NULL){
     library(keras)
     if(is.null(newdata)) newdata = self$data$x
     predict(self$model, newdata)
   },

   serialize = function(){
     keras::serialize_model(model, include_optimizer = TRUE)
   },

   reset_fit = function(){
     self$k$clear_session()
     self$model = keras::keras_model_sequential()
     private$built = FALSE
     self$build()
     self$compile()
   },

   reset_all = function(){
     self$k$clear_session()
     self$model = keras::keras_model_sequential()
     private$built = FALSE
   }


 ),
 private = list(
   architecture = NULL,
   activation_function = NULL,
   bias = NULL,
   regularization = NULL,
   dropout_rate = NULL,
   batch_size = NULL,
   epochs = NULL,
   learning_rate = NULL,
   output_activation = NULL,
   built = NULL
))



#' Base ANN_lm object
#'
#' \code{ANN_lm} object holds an artifical neural network model with a normal distribution likelihood
#'
#' @format \code{\link{R6Class}} object.
#' @name ANN_lm
#' @section Usage:
#' \preformatted{
#' model = ANN$new()
#' }
#'
#' @section Arguments:
#' \describe{
#' \item{model: }{(any)\cr keras model}
#' \item{data: }{(list)\cr List of data with x and y as matrices }
#' \item{data: }{(list)\cr fit history list}
#' \item{build: }{(function)\cr build function}
#'
#'
#' }
#'
#' @section Details:
#' comming soon
#'
#'
#'
#' @author Maximilian Pichler
#' @export
NULL

#'@export
ANN_lm = R6::R6Class("ANN_lm",
  inherit = ANN,
  public = list(

    eps = NULL,
    tf = NULL,
    dist = NULL,
    sigma = NULL,

    initialize = function(...){
      super$initialize(...)
      self$tf = reticulate::import("tensorflow")
      self$dist = reticulate::import("tensorflow_probability")$distributions
      self$init_eps()
    },

    init_eps = function(){
       self$eps = self$tf$constant(.Machine$double.eps)
    },

    build_lm = function(){
      if(self$build()){
        suppressMessages({self$model$layers[[length(self$model$layers)]]$add_weight(name = 'sigma',
                                                                  shape = list(),
                                                                  initializer = keras::initializer_constant(0.5),
                                                                  trainable = TRUE)
        self$eps = self$tf$constant(.Machine$double.eps)
        self$build_output("linear")
        })
      }
    },

    compile = function(){


      normal_likelihood = function(y_true, y_pred){
        sigma = self$tf$get_default_graph()$get_tensor_by_name("sigma:0")
        ll = self$dist$Normal(y_pred, scale = sigma+self$eps)$log_prob(y_true)
        return(self$k$mean(-ll))
      }

      keras::compile(self$model, loss = normal_likelihood, optimizer = keras::optimizer_rmsprop(lr = private$learning_rate))
    },

    train = function(){
      super$train()

      self$update_sigma()
    },

    update_sigma = function() {
      self$sigma = self$k$get_value(self$tf$get_default_graph()$get_tensor_by_name("sigma:0"))
    },

    reset_fit = function(){
       self$k$clear_session()
       self$model = keras::keras_model_sequential()
       private$built = FALSE
       self$build_lm()
       self$compile()
    }



))
